storage/innobase/row/row0query.cc

```diff
@@ -0,0 +1,634 @@
+      break;
+    if (!btr_pcur_move_to_next(&m_pcur, &m_mtr)) break;
+  }
+  m_mtr.commit();
+  return (match_count > 0 || !tuple) ? DB_SUCCESS : DB_RECORD_NOT_FOUND;
+}
+
+dberr_t QueryExecutor::read_by_index(dict_table_t *table,
+                                     dict_index_t *sec_index,
+                                     dtuple_t *search_tuple,
+                                     page_cur_mode_t mode,
+                                     RecordCallback& callback) noexcept
+{
+  ut_ad(table);
+  ut_ad(sec_index);
+  ut_ad(sec_index->table == table);
+  ut_ad(!dict_index_is_clust(sec_index));
+
+  dict_index_t *clust_index= dict_table_get_first_index(table);
+  if (!clust_index) return DB_ERROR;
+
+  m_mtr.start();
+  if (m_trx && !m_trx->read_view.is_open())
+  {
+    trx_start_if_not_started(m_trx, false);
+    m_trx->read_view.open(m_trx);
+  }
+  m_pcur.btr_cur.page_cur.index= sec_index;
+
+  dberr_t err= DB_SUCCESS;
+  if (search_tuple)
+    err= btr_pcur_open(search_tuple, mode, BTR_SEARCH_LEAF, &m_pcur, &m_mtr);
+  else
+  {
+    err= m_pcur.open_leaf(true, sec_index, BTR_SEARCH_LEAF, &m_mtr);
+    if (err == DB_SUCCESS) btr_pcur_move_to_next(&m_pcur, &m_mtr);
+  }
+
+  if (err != DB_SUCCESS)
+  {
+    m_mtr.commit();
+    return err;
+  }
+
+  ulint match_count= 0;
+  while (btr_pcur_is_on_user_rec(&m_pcur))
+  {
+    const rec_t *sec_rec= btr_pcur_get_rec(&m_pcur);
+    rec_offs* sec_offsets= rec_get_offsets(sec_rec, sec_index, nullptr,
+                                           sec_index->n_core_fields,
+                                           ULINT_UNDEFINED, &m_heap);
+    /* Check if secondary record matches our search criteria */
+    RecordCompareAction action= callback.compare_record(search_tuple, sec_rec,
+                                                        sec_index, sec_offsets);
+    if (action == RecordCompareAction::PROCESS)
+    {
+      /* Lookup clustered record and process it */
+      if (!lookup_clustered_record(table, sec_index, clust_index,
+                                   sec_rec, callback, match_count))
+        break;
+    }
+    else if (action == RecordCompareAction::STOP)
+      break;
+    if (!btr_pcur_move_to_next(&m_pcur, &m_mtr)) break;
+  }
+  m_mtr.commit();
+  return (match_count > 0 || !search_tuple) ? DB_SUCCESS : DB_RECORD_NOT_FOUND;
+}
+
+bool QueryExecutor::lookup_clustered_record(dict_table_t *table,
+                                            dict_index_t *sec_index,
+                                            dict_index_t *clust_index,
+                                            const rec_t *sec_rec,
+                                            RecordCallback& callback,
+                                            ulint& match_count) noexcept
+{
+  /* Extract primary key from secondary index record */
+  dtuple_t *clust_tuple= row_build_row_ref(ROW_COPY_DATA, sec_index,
+                                           sec_rec, m_heap);
+  if (!clust_tuple) return false;
```

Comment:
This looks like dead code. If `ref = dtuple_create(heap, ref_len);` returned `nullptr` to `row_build_row_ref()`, the subsequent statement `dict_index_copy_types(ref, clust_index, ref_len);` should crash. So, a `nullptr` return value should be impossible and this should be dead code.